Documenting the supported metrics.

aravindavk opened the pull request:
`docs/metrics.adoc` file is generated using the metrics information
available `docs/metrics.yml` by running `make metrics-docgen`.

Signed-off-by: Aravinda VK <avishwan@redhat.com>

aravindavk commented:
Note to reviewers:

- `docs/metrics.adoc` is generated file. Please review the output format and suggestions related to document format
- `docs/metrics.yml` is the source document where all the metrics details are available.
- If required labels missing in any metrics, we need to add to exporter and to this yml file. 
- Preview is available here https://github.com/gluster/gluster-prometheus/blob/dd1cb63b73f2d381a9b6942455c94bc6dd777bf6/docs/metrics.adoc

JohnStrunk commented on docs/metrics.yml:
```diff
@@ -0,0 +1,137 @@
+
+- name: gluster_memory_percentage
+  description: >
+    Memory percentage of Gluster process. One metric will be exposed
```
100% corresponds to... ?

JohnStrunk commented on docs/metrics.yml:
```diff
@@ -0,0 +1,137 @@
+    volume: Volume name to which the process belongs to.
+
+- name: gluster_brick_capacity_used
+  description: Used capacity of a brick.
```
Need units for capacities as well

shtripat commented:
>> I was thinking how effectively we can avoid document going out of sync. This suggestion helps. Thanks

John, Aravinda,
Can we think of some kind of annotations (specific to our scenario) added to the code and while build we can look for those annotations and generate this document. Just a thought that we can have annotations added at the exporter function level about each of the metrics points it exposes in the form like 

```
// @@@ EXPOSES
// @@@ NAME: gluster_brick_capacity_used
// @@@ DESCRIPTION: Overall brick capacity used
// @@@ LABELS: volume_name, brick_path, ....
// @@@ NAME: gluster_brick_capacity_free
// @@@ DESCRIPTION: Overall brick capacity free
// @@@ LABELS: volume_name, brick_path, ....
func brickUtilization() error {
......
}
```

We can have a script written which while `make metrics-docgen` would read these details from code and generate the doc. This way we dont need to maintain, the separate yml file and every exporter writer needs to write these kind of annotations in the code. If malformed, annotations we can fail the build and dont allow to merge the code in github.

Thoughts???

aravindavk commented:
Added logic to generate documentation from code itself. I am yet to address the comments from @JohnStrunk . I will update the PR with those comments incorporated soon.

aravindavk commented:
> Can we think of some kind of annotations (specific to our scenario) added to the code and while build we can look for those annotations and generate this document. Just a thought that we can have annotations added at the exporter function level about each of the metrics points it exposes in the form like

@shtripat parsing code comments into documentation is hard to maintain. Prometheus already expects Help messages for defining the metrics. I extended the Metrics structure to contain custom values. We can even add tags/category to each metric so that enable/disable support can be added to each tag/category.

shtripat commented:
>> @shtripat parsing code comments into documentation is hard to maintain. Prometheus already expects >> Help messages for defining the metrics. I extended the Metrics structure to contain custom values. We >> can even add tags/category to each metric so that enable/disable support can be added to each
>> tag/category.

Ack

aravindavk commented:
Updated the help messages as suggested. CPU percentage and Memory percentage help messages are gathered from `man ps` (`pcpu` and `pmem`). @JohnStrunk please review

Also changed metric names to include units as suggested by Prometheus naming guidelines(https://prometheus.io/docs/practices/naming/)
